### PR TITLE
Colibread suite

### PR DESCRIPTION
**Pull request by cmonjeau**
Hi,

first PR to include colibread tools in IUC (old versions with old packages are in the main toolshed).

tool 1/6 -> kissplice
tool 2/6 -> discosnp++
tool 3/6 -> mapsembler2
tool 4/6 -> commet
tool 5/6 -> takeabreak
tool 6/6 -> Lordec

### CODE REVIEW

File: tools/colibread/discosnp_pp.xml [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        #for $input in $list_reads
+            #if $input
+                #set $filename = str($input)
+                #if $input.is_of_type('fasta')
+                    #set $filename = $filename + ".fasta"
+                #else if $input_type.is_of_type('fastq')
+                    #set $filename = $filename + ".fastq"
+                #end if
+                ln -sf '${input}' '${filename}' &&
+                echo "${filename}" >> input.lst &&
```
Comment (bgruening) [inline tools/colibread/discosnp_pp.xml]: single quotes

File: tools/colibread/discosnp_pp.xml [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            #end if
+        #end for
+
+        ## if paired reads in a paired:list
+        #set $count = 0
+        #for $paired in $list_paired_reads
```
Comment (bgruening) [inline tools/colibread/discosnp_pp.xml]: you can also use `enumerate()` here.

File: tools/colibread/discosnp_pp.xml [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+            #set $filenameREV = str($paired.reverse)
+            #if $paired.reverse.is_of_type('fasta')
+                #set $filenameREV = $filenameREV + ".fasta"
+            #else if $paired.reverse.is_of_type('fastq')
+                #set $filenameREV = $filenameREV + ".fastq"
+            #end if
+            ln -sf '${paired.reverse}' '${filenameREV}' &&
+
+            echo "${filenameFWD}" > "paired_${count}.lst" &&
```
Comment (bgruening) [inline tools/colibread/discosnp_pp.xml]: Do you need the `paired_${count}.lst` files later? Why not directly putting the content into input.lst?

File: tools/colibread/discosnp_pp.xml [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            -G '${reference_file}'
+            -M ${VCF_option.M}
+            ${VCF_option.R}
+        #end if
+
+  ]]></command>
+
+  <inputs>
+
+      <param name='list_reads' format="fasta,fastq" type="data" multiple="true" label="Single read files" optional="True" />
```
Comment (bgruening) [inline tools/colibread/discosnp_pp.xml]: shouldn't this be a conditional like here: https://github.com/galaxyproject/tools-iuc/pull/1500/files#diff-d9d476791f89bdc57b8b237e3e263389R47

Both are currently optional, so no input is required?

File: tools/colibread/discosnp_pp.xml [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        #end for
+
+        ## if paired reads in a paired:list
+        #set $count = 0
+        #for $paired in $list_paired_reads
+
+            #set $count += 1
+
+            #set $filenameFWD = str($paired.forward)
+            #if $paired.forward.is_of_type('fasta')
```
Comment (bgruening) [inline tools/colibread/discosnp_pp.xml]: Just a thought that needs to be tested, but I guess you can simplify this a lot with `#set $filenameFWD = $filenameFWD + $filenameFWD.ext`

File: tools/colibread/kissplice.xml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        $samples
+        -k ${k}
+        -c ${c}
+        -C ${C}
+        ${output_context}
+        --counts ${counts}
+
+    ]]></command>
+    <inputs>
+        <conditional name="input_type">
```
Comment (bgruening) [inline tools/colibread/kissplice.xml]: We could save this entire conditional if you make the simple_read `multiple=True` isn't it?

File: tools/colibread/kissplice.xml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        <param argument="-C" type="float" value="0.05" label="Edges with relative coverage below this number are removed (value in [0,1) required)" />
+        <param name="output_context" argument="--output-context" type="boolean" checked="false" truevalue="--output-context" falsevalue="" label="Output the maximum non-ambiguous context of each event" />
+        <param argument="--counts" type="select" label="Selecting how the counts will be reported">
+            <option value="0">Total counts</option>
+            <option value="1">Counts on junctions</option>
+            <option value="2">All counts</option>
+        </param>
+    </inputs>
+    <outputs>
+        <data name="type0a" format="fasta" label="Single SNPs or sequencing errors" from_work_dir="results/*_coherents_type_0a.fa"/>
```
Comment (bgruening) [inline tools/colibread/kissplice.xml]: Are all outputs needed here? Would it make sense to offer a user-selectable output?

File: tools/colibread/discosnp_pp.xml (hunk 1 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: oops

File: tools/colibread/discosnp_pp.xml (hunk 2 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: done

File: tools/colibread/discosnp_pp.xml (hunk 5 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: I use this kind of line :  #set $filename = str($x) + "." + $x.ext

File: tools/colibread/discosnp_pp.xml (hunk 3 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: Yes. Input is a file of files. I can't put all paired-end reads in a unique file. I need one file per paired-end reads. It's the software need.

File: tools/colibread/discosnp_pp.xml (hunk 4 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: User can use one of the two inputs or both. I didn't found a valid solution. Any ideas?

File: tools/colibread/kissplice.xml (hunk 6 above)
Comment (cmonjeau) [reply tools/colibread/kissplice.xml]: Yeah

**Comment by cmonjeau** [comment]
Thanks for this first review @bgruening :)

File: tools/colibread/discosnp_pp.xml [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <requirements>
+        <requirement type="package" version="2.2.10">discosnp</requirement>
+        <requirement type="package" version="0.7.16">bwa</requirement>
+    </requirements>
+    <command><![CDATA[
+        ##create an empty input file
+        touch input.lst &&
```
[the remark this answers is not in the document]
Comment (bgruening) [reply tools/colibread/discosnp_pp.xml]: ping

File: tools/colibread/discosnp_pp.xml (hunk 5 above)
Comment (bgruening) [reply tools/colibread/discosnp_pp.xml]: Have you tested it? Maybe it does not work if your filetype is `fastqsanger`

File: tools/colibread/discosnp_pp.xml [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+            #set $filenameFWD = str($paired.forward) + "." + $paired.forward.ext
+            ln -sf '${paired.forward}' '${filenameFWD}' &&
+
+            #set $filenameREV = str($paired.reverse) + "." + $paired.reverse.ext
+            ln -sf '${paired.reverse}' '${filenameREV}' &&
+
+            echo '${filenameFWD}' > "paired_${count}.lst" &&
+            echo '${filenameREV}' >> "paired_${count}.lst" &&
+            echo "paired_${i}.lst" >> input.lst &&
```
Comment (bgruening) [inline tools/colibread/discosnp_pp.xml]: are the first two echos needed?

File: tools/colibread/discosnp_pp.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: Deleted 👍

File: tools/colibread/discosnp_pp.xml (hunk 4 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: I use a conditional now!

File: tools/colibread/kissplice.xml (hunk 7 above)
Comment (cmonjeau) [reply tools/colibread/kissplice.xml]: I added a select for that.

File: tools/colibread/discosnp_pp.xml (hunk 9 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: Unfortunately yes. Discosnp need this kind of input structure -> https://github.com/GATB/DiscoSnp/blob/master/doc/discoSnp_user_guide.pdf

File: tools/colibread/discosnp_pp.xml (hunk 5 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: It seems that tools check if file extension is .fastq* so fastqsanger extension works.

File: tools/colibread/commet.xml [hunk 10]
```diff
@@ -0,0 +1,109 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool profile="16.06" id="commet" name="commet" version="24.7.14">
```
[the remark this answers is not in the document]
Comment (abretaud) [reply tools/colibread/commet.xml]: some capital letter for the tool name would look cleaner?

File: tools/colibread/commet.xml [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+    <inputs>
+        <!-- Input data files -->
+        <repeat name="sets" title="Read sets" min="1">
+            <param name="reads" type="data" multiple="true" format="fasta, fastq" label="Dataset" help="Accept fasta/fastq"/>
+        </repeat>
+        <param argument="-k" type="integer" label="Size of kmers" value="33" help="Set the length of used kmers." />
+        <param argument="-t" type="integer" label="Mini shared kmers" value="2" help="Minimal number of shared kmers." />
+        <param argument="-l" type="integer" label="Read mini length" value="66" help="Minimal length a read should have to be kept." />
+        <param argument="-e" type="float" label="Mini Shannon index" value="0" help="Minimal Shannon index a read should have to be kept. Float in [0,2.32]" />
```
Comment (abretaud) [inline tools/colibread/commet.xml]: can you add min and max?

File: tools/colibread/discosnp_pp.xml [hunk 12]
```diff
@@ -0,0 +1,200 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool profile="16.04" id="discosnp_pp" name="discoSnp++" version="2.2.4">
+    <description>is an efficient tool for detecting SNPs without a reference genome.</description>
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <requirements>
+        <requirement type="package" version="2.2.10">discosnp</requirement>
```
Comment (abretaud) [inline tools/colibread/discosnp_pp.xml]: 2.2.10 doesn't match 2.2.4 on line 2

File: tools/colibread/discosnp_pp.xml [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        <requirement type="package" version="0.7.16">bwa</requirement>
```
Comment (abretaud) [inline tools/colibread/discosnp_pp.xml]: Shouldn't bwa be a requirement in discosnp conda recipe?

File: tools/colibread/discosnp_pp.xml [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+              <param name='list_paired_reads' argument="-r" format="fasta,fastq" type="data_collection" collection_type="list:paired" multiple='true' label="List of paired read files" />
+          </when>
+          <when value="mix">
+              <param name='list_reads' argument="-r" format="fasta,fastq" type="data" multiple="true" label="Single read files" />
+              <param name='list_paired_reads' argument="-r" format="fasta,fastq" type="data_collection" collection_type="list:paired" multiple='true' label="List of paired read files"/>
+          </when>
+      </conditional>
+
+      <param argument="-b" type="select" label="Branching strategy">
+          <option value="0">variants for which any of the two paths is branching are discarded</option>
```
Comment (abretaud) [inline tools/colibread/discosnp_pp.xml]: Capitalize first letters of all options

File: tools/colibread/discosnp_pp.xml [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+              <param argument="-M" type="integer" value="4" label="Maximal number of mapping errors" help="during BWA mapping phase" />
+              <param argument="-R" type="boolean" truevalue="-R" falsevalue="" checked="false" label="Use the reference genome also in the variant calling, not only for mapping results" />
+          </when>
+      </conditional>
+
+  </inputs>
+
+  <outputs>
+      <data name="vcf" from_work_dir="*_coherent.vcf" format="vcf" label="VCF with ${tool.name} on $on_string"/>
+      <data name="fasta" from_work_dir="*_coherent.fa" format="fasta" label="Multifasta with ${tool.name} on $on_string"/>
```
Comment (abretaud) [inline tools/colibread/discosnp_pp.xml]: Is it just one file with an unpredictable name?

File: tools/colibread/kissplice.xml [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    <version_command>kissplice --version</version_command>
+    <command><![CDATA[
+
+        ## prepare symlinks
+        #set $samples = ""
+
+        #for x in $list_reads
+            #set $filename = str($x) + "." + $x.ext
+            ln -sf '${x}' '${filename}' &&
+            #set $samples += " -r " + $filename
```
Comment (abretaud) [inline tools/colibread/kissplice.xml]: I think you need to add single quotes on $filename here

File: tools/colibread/kissplice.xml [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+References, documentation and FAQ here_.
+
+.. _here: http://kissplice.prabi.fr/
+
+
+    ]]></help>
+    <expand macro="citations">
+        <citation type="bibtex">
+         @article{PMID:22537044,
```
Comment (abretaud) [inline tools/colibread/kissplice.xml]: Can you use the doi instead?

File: tools/colibread/lordec.xml [hunk 18]
```diff
@@ -0,0 +1,150 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool profile="16.06" id="lordec" name="lordec" version="0.6">
```
Comment (abretaud) [inline tools/colibread/lordec.xml]: 16.06

Comment (abretaud) [reply tools/colibread/lordec.xml]: "Lordec" for name?

File: tools/colibread/mapsembler2.xml [hunk 19]
```diff
@@ -0,0 +1,110 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool profile="16.06" id="mapsembler2" name="Mapsembler2" version="2.2.4">
```
Comment (abretaud) [inline tools/colibread/mapsembler2.xml]: 16.06

File: tools/colibread/mapsembler2.xml [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+    <inputs>
+        <!-- Input data files -->
+        <param argument="-s" type="data" format="fasta" label="Starters" help="set of input sequences" />
+        <param argument="-r" type="data" multiple="true" format="fasta,fastq" label="List of reads" />
+        <param argument="-t" type="select" label="Select your output extension type">
+            <option value="1">a strict sequence</option>
+            <option value="2">a consensus sequence</option>
+        </param>
+        <param argument="-k" type="integer" label="Size of kmers" value="31" help="Set the length of used kmers. Must fit the compiled value. Only uneven number" />
+        <param argument="-c" type="integer" label="Minimal coverage" value="5" help="set the minimal coverage: Used by Phaser (don't use kmers with lower coverage) "/>
```
Comment (abretaud) [inline tools/colibread/mapsembler2.xml]: For -c and -d capitalize first letter of label

File: tools/colibread/takeabreak.xml [hunk 21]
```diff
@@ -0,0 +1,121 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool profile="16.06" id="takeabreak" name="TakeABreak" version="1.1.2">
```
Comment (abretaud) [inline tools/colibread/takeabreak.xml]: 16.04

File: tools/colibread/takeabreak.xml [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+                <param name='reads' argument="-in" format="fasta,fastq" type="data" multiple="true" label="Single read files" optional="True" />
+            </when>
+            <when value="list">
+                <repeat name="reads_lists" title="Individual" min="1">
+                    <param name='list_reads' argument="-in" format="fasta,fastq" type="data_collection" collection_type="list" multiple="true" label="Individual read files list"/>
+                </repeat>
+            </when>
+        </conditional>
+        <param name="kmer_size" argument="-kmer-size" type="integer" label="length of the k-mers" value="31" help="Set the length of used kmers" />
+        <param name="abundance_min" argument="-abundance-min" type="text" label="the minimal abundance threshold" value="auto" />
```
Comment (abretaud) [inline tools/colibread/takeabreak.xml]: Isn't it an integer?

File: tools/colibread/commet.xml (hunk 11 above)
Comment (cmonjeau) [reply tools/colibread/commet.xml]: Software documentation doesn't mention any min or max for options.

File: tools/colibread/discosnp_pp.xml (hunk 12 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: Fix

File: tools/colibread/discosnp_pp.xml (hunk 14 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: Ok

File: tools/colibread/discosnp_pp.xml (hunk 15 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: The name is (almost) unpredictable. It's a long name composed by user parameters and underscores and letters.

File: tools/colibread/discosnp_pp.xml (hunk 13 above)
Comment (cmonjeau) [reply tools/colibread/discosnp_pp.xml]: Indeed. Recipe has been corrected.

File: tools/colibread/kissplice.xml (hunk 16 above)
Comment (cmonjeau) [reply tools/colibread/kissplice.xml]: Mmmm where? I can't add single quotes in the #set line

Comment (abretaud) [reply tools/colibread/kissplice.xml]: Something like this?
`#set $samples += " -r '$filename'"`

File: tools/colibread/kissplice.xml (hunk 17 above)
Comment (cmonjeau) [reply tools/colibread/kissplice.xml]: doi is dead :/

File: tools/colibread/lordec.xml (hunk 18 above)
Comment (cmonjeau) [reply tools/colibread/lordec.xml]: ok (maiiiiilll)

File: tools/colibread/mapsembler2.xml (hunk 19 above)
Comment (cmonjeau) [reply tools/colibread/mapsembler2.xml]: Ok

File: tools/colibread/mapsembler2.xml (hunk 20 above)
Comment (cmonjeau) [reply tools/colibread/mapsembler2.xml]: Done

File: tools/colibread/takeabreak.xml (hunk 21 above)
Comment (cmonjeau) [reply tools/colibread/takeabreak.xml]: oops

File: tools/colibread/takeabreak.xml (hunk 22 above)
Comment (cmonjeau) [reply tools/colibread/takeabreak.xml]: Yes and No. The default value is auto. If there are several inputs, value can be a list of integers.

File: tools/colibread/kissplice.xml (hunk 16 above)
Comment (cmonjeau) [reply tools/colibread/kissplice.xml]: Does not work. Command line is `kissplice -r '$filename' -r '$filename'`

Comment (abretaud) [reply tools/colibread/kissplice.xml]: What about `#set $samples += " -r '" + $filename + "'"`?

Comment (cmonjeau) [reply tools/colibread/kissplice.xml]: Yes it works. But is it necessary? The path is generated by galaxy so normally it's good!

Anyway, thanks Mr La Bretouille

File: tools/colibread/kissplice.xml (hunk 17 above)
Comment (bgruening) [reply tools/colibread/kissplice.xml]: Indeed https://doi.org/10.1186/1471-2105-13-S6-S5 seems to be dead.

**Comment by bgruening** [comment]
@abretaud ready if you are.

**Comment by abretaud** [comment]
I am! Thank you @cmonjeau 
you're awesome

**Pull request merged**